onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -920,28 +1028,41 @@ template common::Status GetSizeInBytesFromTensorProto<kAllocAlignment>(const ONN
-    } else {                                                                    \
-      element_count = initializer.DATA_SIZE();                                  \
-      tensor_byte_size = element_count * sizeof(ELEMENT_TYPE);                  \
-    }                                                                           \
-    unpacked_tensor.reset(new uint8_t[tensor_byte_size]);                       \
-    return onnxruntime::utils::UnpackTensor(                                    \
-        initializer,                                                            \
-        initializer.has_raw_data() ? initializer.raw_data().data() : nullptr,   \
-        initializer.has_raw_data() ? initializer.raw_data().size() : 0,         \
-        reinterpret_cast<ELEMENT_TYPE*>(unpacked_tensor.get()), element_count); \
-    break;                                                                      \
+#define CASE_UNPACK(TYPE, ELEMENT_TYPE, DATA_SIZE)                                         \
+  case ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_##TYPE: {                \
+    if (initializer.data_location() == TensorProto_DataLocation_EXTERNAL) {                \
+      ORT_RETURN_IF_ERROR(ReadExternalDataForTensor(                                       \
+          initializer,                                                                     \
+          model_path.IsEmpty() ? nullptr : model_path.ParentPath().ToPathString().c_str(), \
+          unpacked_tensor,                                                                 \
+          tensor_byte_size));                                                              \
+      tensor_data_length = tensor_byte_size;                                               \
+      return Status::OK();                                                                 \
+    } else {                                                                               \
+      size_t element_count = 0;                                                            \
+      if (initializer.has_raw_data()) {                                                    \
+        tensor_byte_size = initializer.raw_data().size();                                  \
+        element_count = tensor_byte_size / sizeof(ELEMENT_TYPE);                           \
+      } else {                                                                             \
+        element_count = initializer.DATA_SIZE();                                           \
+        tensor_byte_size = element_count * sizeof(ELEMENT_TYPE);                           \
+      }                                                                                    \
+      tensor_data_length = tensor_byte_size;                                               \
+      unpacked_tensor.reset(new uint8_t[tensor_byte_size]);                                \
+      return onnxruntime::utils::UnpackTensor(                                             \
```

Comment:
Can just use tensor_data_length here as that contains the size_t from tensor_byte_size. Otherwise you need `*&tensor_byte_size`